R/SEraster.R

```diff
@@ -0,0 +1,492 @@
+#' @keyword internal
+#' 
+createRasterizedObject <- function(input, out, name) {
+  data_rast <- out$data_rast
+  pos_rast_temp <- as.data.frame(out$pos_rast)
+  
+  if(name %in% Assays(input)){
+    image <- Images(input, assay = name)[1]
+  } else{
+    image <- Images(input, assay = DefaultAssay(input))[1]
+  }
+  
+  meta_rast <- out$meta_rast[,c("num_cell","type","resolution")]
+  resolution <- meta_rast[["resolution"]][1]
+  for (i in seq_along(out$meta_rast$cellID_list)) {
+    meta_rast$cellID_list[i] <- paste(unlist(out$meta_rast$cellID_list[[i]]), collapse = ", ")
+  }
+  
+  output <- CreateSeuratObject(
+    counts = data_rast,
+    assay = name,
+    meta.data = meta_rast
+  )
+  
+  class <- class(input[[image]])
+  if(class == 'VisiumV1') {
+    scale.use <- ScaleFactors(input[[image]])[['lowres']]
+    pos_rast <- as.data.frame(cbind(tissue = rep(1, nrow(pos_rast_temp)), 
+                                     row = pos_rast_temp$x, 
+                                     col = pos_rast_temp$y,
+                                     imagerow = ceiling(pos_rast_temp$x / scale.use),
+                                     imagecol = ceiling(pos_rast_temp$y / scale.use)))
+    rownames(pos_rast) <- rownames(pos_rast_temp)
+    input[[image]]@scale.factors$spot <- sqrt(nrow(input[[]])/nrow(meta_rast))*Radius(input[[image]])
+    
+    new.fov <- new(
+      Class = "VisiumV1",
+      coordinates = pos_rast,
+      assay = name, 
+      key = "rasterized_",
+      image = input[[image]]@image,
+      scale.factors = input[[image]]@scale.factors,
+      spot.radius = input[[image]]@scale.factors$spot
+    )
+  } else if(class == 'VisiumV2') {
+    input[[image]]@scale.factors$spot <- sqrt(nrow(input[[]])/nrow(meta_rast))*slot(input[[image]]$centroids, name='radius')
+    fov <- CreateFOV(
+      pos_rast_temp[, c("x", "y")],
+      type = "centroids",
+      radius = input[[image]]@scale.factors[["spot"]],
+      assay = name,
+      key = Key("rasterized_", quiet = TRUE)
+    )
+    new.fov <- new(
+      Class = "VisiumV2",
+      boundaries = fov@boundaries,
+      assay = fov@assay,
+      key = fov@key,
+      image = input[[image]]@image,
+      scale.factors = input[[image]]@scale.factors
+    )
+  } else if(class == 'FOV') {
+    radius <- sqrt(nrow(input[[]])/nrow(meta_rast))*slot(input[[image]]$centroids, name='radius')
+    pos_rast_temp$cell <- rownames(pos_rast_temp)
+    segmentations.data <- list(
+      "centroids" = CreateCentroids(coords = pos_rast_temp[, c("x", "y")], nsides = 4, radius = radius),
+      "segmentation" = CreateSegmentation(coords = pos_rast_temp)
+    )
+    new.fov <- CreateFOV(
+      coords = segmentations.data,
+      type = c("segmentation", "centroids"),
+      radius = radius,
+      molecules = input[[image]]$molecules,
+      assay = name,
+      key = Key("rasterized_", quiet = TRUE)
+    )
+  }
+  output@images[[paste0(name,".rasterized")]]<- new.fov
+  return(output)
+}
```

Comment:
The reason I didn't do this was because existing VisiumV1 class objects (like stxbrain in SeuratData) don't have Centroids, and VisiumV2 objects don't have segmentations because that's only relevant to in-situ data, so you wouldn't be able to extract info with 

```
input_fov <- input[[image_name]]
input_centroids <- input.fov[["centroids"]]
input_segmentation <- input.fov[["segmentation"]]
input_molecules <- input.fov[["molecules"]]
```

Not sure if there is really an alternative but if you have other ideas please let me know!

Reply:
It should be pretty easy to essentially ignore the "segmentation"s and pass them straight through:

```R
boundary_names <- Boundaries(input_fov)
input_boundaries <- input_fov@boundaries
input_centroids <- ifelse(
    "centroids" %in% boundary_names,
    input_boundaries[["centroids"]],
    NULL
)
input_molecules <- input_fov@molecules
```

Wrapping the whole thing in a `tryCatch` allows us to accomodate non-`FOV` images:

```R
tryCatch(
    {
      boundary_names <- Boundaries(input_fov)
      input_boundaries <- input_fov@boundaries
      input_centroids <- ifelse(
        "centroids" %in% boundary_names,
        input_boundaries[["centroids"]],
        NULL
      )
      input_molecules <- input_fov@molecules
    },
    error = function(e) {
      input_boundaries <- list()
      input_centroids <- NULL
      input_molecules <- NULL
    }
  )
```

without having to know the actual type being handed around:

```R
if (!is.null(input_centroids)) {
    output_nsides <- input_centroids@nsides
    output_theta <- input_centroids@theta

    output_radius <- sqrt(
      nrow(input[[]]) / nrow(meta_rast)
    ) * Radius(input_centroids) # unscaled
  } else {
    nsoutput_nsidesides <- NULL
    output_theta <- NULL

    output_radius <- Radius(input_fov)
  }

  output_centroids <- CreateCentroids(
    coords = output_coordinates,
    nsides = output_nsides,
    radius = output_radius,
    theta = output_theta
  )
```

I guess you could technically also get an `FOV` with _multiple_ `Molecules` but all that means is we should avoid using `CreateFOV` and just call `new` directly (also note the way "segmentations" will be handed through if it's present):

```R
output_boundaries <- input_boundaries
  output_boundaries[["centroids"]] <- output_centroids

  output_fov <- new(
    Class = "FOV",
    boundaries = output_boundaries,
    molecules = input_molecules,
    assay = assay,
    key = Key(output_image_name, quiet = TRUE)
  )
```